rst_files/mccall_model_with_separation.rst

```diff
@@ -216,55 +216,51 @@ Implementation
 
 Let's implement this iterative process
 
-In the code you'll see that we use a type to store the various parameters and other
-objects associated with a given model
-
-This helps to tidy up the code and provides an object that's easy to pass to functions
-
-The default utility function is a CRRA utility function
-
 .. code-block:: julia
     :class: test
 
     using Test
 
-.. code-block:: julia
+.. code-block:: julia 
 
     using Distributions, LinearAlgebra, Compat, Expectations, Parameters, NLsolve, Plots
 
-    # a default utility function
-    u(c, σ) = (c^(1 - σ) - 1) / (1 - σ) + (c <= 0)*(-10e6) # if c > 0, return utility; else, return large negative
-
-    # model objects
-    const n = 60                                           # n possible outcomes for wage
-    const default_w_vec = range(10, 20, length = n)        # wages between 10 and 20
-    const a, b = 600, 400                                  # shape parameters
-    const dist = BetaBinomial(n-1, a, b)                   # wage distribution
-
-    # constructor
-    McCallModel(;α = 0.2, β = 0.98, γ = 0.7, c = 6.0, σ = 2.0, w_vec = default_w_vec, dist = dist) = (α = α, β = β, γ = γ, c = c, σ = σ, w_vec = w_vec, dist = dist)
-
-    # update
-    function update_bellman(mcm, V, E)
-        @unpack α, β, σ, c, γ, dist, w_vec = mcm # unpack model objects
-        U = V[end] # U is at the end of V
-        V_new = similar(V) # create a new V
-
-        for (w_idx, w) in enumerate(mcm.w_vec)
-            # w_idx indexes the vector of possible wages
-            V_new[w_idx] = u(w, σ) + β * ((1 - α) * V[w_idx] + α * U)
+    function solve_mccall_model(mcm; U_iv = 1.0, V_iv = ones(length(mcm.w)), tol = 1e-5, iter = 2_000)
+        # α, β, σ, c, γ, w = mcm.α, mcm.β, mcm.σ, mcm.c, mcm.γ, mcm.w
+        @unpack α, β, σ, c, γ, w, dist, u = mcm 
+        
+        # parameter validation
+        @assert c > 0.0
+        @assert minimum(w) > 0.0 # perhaps not strictly necessary, but useful here 
+        
+        # necessary objects 
+        u_w = u.(w, σ) 
+        u_c = u(c, σ)
+        E = expectation(dist) # expectation operator for wage distribution
+
+        # Bellman operator T. Fixed point is x* s.t. T(x*) = x*
```

Comment:
Why not use a more descriptive name than `T` unless used as such in the math?

Reply:
Hopefully the math had T or something similar. If not, we could add the operator in to make it clear.. Usually it was called T in other lectures?

Reply:
I think we should add the operator to the math on this one. But @Nosferican, I also think the `x = Tx` convention for the iteration is commonplace enough that it should be clear what this is doing.